pppd/lcp.c

```diff
@@ -2225,6 +2240,61 @@ LcpEchoTimeout (void *arg)
     }
 }
 
+/*
+ * lcp_rtt_update_buffer - log the RTT of the received LCP echo-request
+ *
+ * The header section at the beginning of lcp_rtt_file contains
+ * LCP_RTT_HEADER_LENGTH fields, each a u_int32_t in network byte order:
+ * [0] LCP_RTT_MAGIC
+ * [1] status (1: the file is open and is being written)
+ * [2] index of the most recently updated element
+ * [3] the value of the lcp-echo-interval parameter
+ *
+ * The header is followed by a ring buffer of LCP_RTT_ELEMENTS elements, each
+ * containing a pair of u_int32_t in network byte order with this content:
+ * [0] UNIX timestamp
+ * [1] bits 24-31: the number of lost LCP echo replies
+ *     bits 0-23:  the measured RTT in microseconds
+ *
+ * Consumers of lcp_rtt_file are expected to:
+ * - read the complete file of arbitrary lenght
+ * - check the magic number
+ * - process the data elements starting at the index
+ * - discard the last data element, i.e. the one preceding the element
+ *   referenced by the index
+ * - ignore any elements with a timestamp of 0
+ */
+static void
+lcp_rtt_update_buffer (unsigned long rtt)
+{
+    volatile u_int32_t *const ring_buffer = lcp_rtt_buffer
+	+ LCP_RTT_HEADER_LENGTH;
+    unsigned int next_entry, lost;
+
+    /* choose the next entry where the data will be stored */
+    if (ntohl(lcp_rtt_buffer[2]) >= (LCP_RTT_ELEMENTS - 1) * 2)
+	next_entry = 0;				/* go back to the beginning */
+    else
+	next_entry = ntohl(lcp_rtt_buffer[2]) + 2;	/* use the next one */
+
+    /* update the data element */
+    /* storing the timestamp in an *unsigned* long allows dates up to 2106 */
+    ring_buffer[next_entry] = htonl((u_int32_t) time(NULL));
+    lost = lcp_echos_pending - 1;
+    if (lost > 0xFF)
+	lost &= 0xFF;		/* truncate the lost packets to 256 */
+    if (rtt > 0xFFFFFF)
+	rtt &= 0xFFFFFF;	/* truncate the RTT to 16777216 */
+    /* use bits 24-31 for the lost packets and bits 0-23 for the RTT */
+    ring_buffer[next_entry + 1] = htonl((u_int32_t) ((lost << 24) + rtt));
+
+    /* update the pointer to the (just updated) most current data element */
+    lcp_rtt_buffer[2] = htonl(next_entry);
```

Comment:
Normally with a mmaped ring buffer one should really have some sort of memory barrier here (i.e. before the write to lcp_rtt_buffer[2]), since otherwise, on architectures with a weakly-consistent memory model, a consumer that has mmaped the ring buffer might see the write to lcp_rtt_buffer[2] before the write to the ring buffer entry. If all consumers are simply going to read the whole buffer with the read() syscall, and we assume that read() is atomic with respect to msync(), then perhaps there isn't actually a problem. I don't know of portable constructs for memory barriers short of using a pthread mutex, and arguably this data is not so critical that an occasional bad value is not a real problem, so perhaps it is reasonable not to have any barrier. But at least a comment may be useful.

Reply:
That was my reasoning as well. I have added a comment.

If this will show to not be true in practice then I have thought of also setting the timestamp of the oldest sample to zero (i.e. the one which would be overwritten the next cycle): this way even if the index is read out of order then no wrong sample is read and the correct one will still be picked up by the reader the next time.